dataframe.py

```diff
@@ -0,0 +1,270 @@
+        return "string"
+
+
+class Object(PrimitiveType):
+    """
+    Any PyObject value
+    """
+    def to_string(self):
+        return "object"
+
+
+class Categorical(DataType):
+    """
+    A categorical value is an ordinal (integer) value that references a
+    sequence of category values of an arbitrary data type
+    """
+
+    def __init__(self, index_type: IntegerType, category_type: DataType,
+                 ordered: bool = False):
+        self.index_type = index_type
+        self.category_type = category_type
+        self.ordered = ordered
+
+    def equals(self, other: DataType) -> bool:
+        return (isinstance(other, Categorical) and
+                self.index_type == other.index_type and
+                self.category_type == other.category_type and
+                self.ordered == other.ordered)
+
+    def to_string(self):
+        return ("categorical(indices={}, categories={}, ordered={})"
+                .format(str(self.index_type), str(self.category_type),
+                        self.ordered))
+
+
+# ----------------------------------------------------------------------
+# Classes representing a column in a DataFrame
+
+
+class Column(ABC):
+
+    @property
+    @abstractmethod
+    def name(self) -> Hashable:
+        pass
+
+    @property
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    def to_numpy(self):
+        """
+        Access column's data as a NumPy array. Recommended to return a view if
+        able but not required
+        """
+        raise NotImplementedError("Conversion to NumPy not available")
+
+    def to_arrow(self, **kwargs):
+        """
+        Access column's data in the Apache Arrow format as pyarrow.Array or
+        ChunkedArray. Recommended to return a view if able but not required
+        """
+        raise NotImplementedError("Conversion to Arrow not available")
+
+
+# ----------------------------------------------------------------------
+# DataFrame: the main public API
+
+
+class DataFrame(ABC, Mapping):
+    """
+    An abstract data frame base class.
+
+    A "data frame" represents an ordered collection of named columns. A
+    column's "name" is permitted to be any hashable Python value, but strings
+    are common. Names are not required to be unique. Columns may be accessed by
+    name (when the name is unique) or by position.
+    """
+
+    def __dataframe__(self):
+        """
+        Idempotence of data frame protocol
+        """
+        return self
+
+    def __iter__(self):
+        # TBD: Decide what iterating should return
+        return iter(self.column_names)
+
+    def __len__(self):
+        return self.num_rows
+
+    @property
+    @abstractmethod
+    def num_columns(self):
+        """
+        Return the number of columns in the DataFrame
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def num_rows(self):
```

Comment:
Same, I don't think we need an alias for `len(df)`, I think using `len()` is the Pythonic way, and it's enough.

Reply:
There's a certain ambiguity to `len(df)`, though. Because the length of a dict/mapping is the number of keys that it has (which here is the number of columns). We take for granted from pandas that `len(df)` returns the number of rows

In R

```R
> df <- data.frame(a=c(1,2,3,4), b=c(5,6,7,8))
> length(df)
[1] 2
> nrow(df)
[1] 4
> ncol(df)
[1] 2
```

Reply:
I agree on that.

To me it makes more sense that `len(df)` is the number of rows, and not of columns. And I'd avoid making a dataframe behave like a dict of its columns. But that's just my preference, and of course that's a reasonable choice.

But my point here was more to have just one way of doing things, and not repeat `num_rows` or `num_columns` for what can be done with `len`. I think we should follow the recommendation from the zen of Python `There should be one-- and preferably only one --obvious way to do it.`. Makes things simple for users IMO.

Reply:
I agree with @wesm here that `len(df)` is ambiguous and prefer the explicit `num_rows` property.

Reply:
IMHO `len` should denote the length of `list(df)` so you have to decide if it's an iterable of rows or columns.

...or just punt on it, let `len(df)` and `iter(df)` raise and make the properties  `rows(self) -> Iterable[Row]` and `columns(self) -> Iterable[Column]` return iterables (which themselves implement `__len__`)

Reply:
Coming back to this, couple of reasons why `len(df)` can not be appropriate (besides some people finding it ambiguous on which dimension it should return).

CPython limits the dunder method `__len__` to an int64:
```python
Python 3.7.6 | packaged by conda-forge | (default, Jun  1 2020, 18:57:50) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class LenTooBig:
...     def __len__(self):
...         return 2 ** 64 + 1
... 
>>> len(LenTooBig())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: cannot fit 'int' into an index-sized integer
```
That is many trillions of data points, that not sure if anyone will ever reach (surely not in-memory representations), but still worth mentioning that the limitation doesn't exist in regular methods.

Another thing, with distributed systems in mind, does it make sense that this is a property (or `__len__`)? Personally, if I see:
```python
df.num_rows  # also applies to `len(df)`
```
I'm mostly assuming that `num_rows` is known / easily available. But for distributed systems that can be an expensive operation. Would it make more sense something like `df.count_rows()`? Also, I personally find more obvious to see that `df.count_rows()` can fail (and return an exception), than `df.num_rows`, which I'd expect to always work, be fast, and not consume significant resources.

Reply:
Making it a function makes sense to me for the reasons you listed. It's probably best in this utilitarian interface to not have `__len__`